File: amidisk_cpp/src/rdb/rescue.cpp
#include "rescue.h"

namespace amidisk {

std::string Candidate::describe() const {
    char buf[128];
    std::string exact_str = exact ? "" : "  (size estimated)";
    std::string lbl = label.empty() ? "-" : ("'" + label + "'");
    snprintf(buf, sizeof(buf), "%-5s @sector %-9llu %8.1f MB  %-7s %s%s",
             source.c_str(), (unsigned long long)start_sec, (num_secs * 512) / 1e6,
             dos_type_to_str(dos_type).c_str(), lbl.c_str(), exact_str.c_str());
    return std::string(buf);
}

static bool ffs_root_valid(const std::vector<uint8_t>& raw, uint32_t bs) {
    if (raw.size() < bs) return false;
    uint32_t type = (raw[0] << 24) | (raw[1] << 16) | (raw[2] << 8) | raw[3];
    if (type != 2) return false;
    
    int32_t ext = (raw[bs-4] << 24) | (raw[bs-3] << 16) | (raw[bs-2] << 8) | raw[bs-1];
    if (ext != 1) return false;
    
    uint32_t s = 0;
    for (uint32_t i = 0; i < bs; i += 4) {
        uint32_t v = (raw[i] << 24) | (raw[i+1] << 16) | (raw[i+2] << 8) | raw[i+3];
        s += v;
    }
    return s == 0;
}

std::pair<std::vector<Candidate>, std::vector<std::string>> Rescue::scan(std::shared_ptr<BlockDevice> blkdev, uint64_t rdb_area_secs, std::function<void(uint64_t, uint64_t)> progress) {
    uint64_t total = blkdev->num_blocks();
    if (rdb_area_secs == 0) rdb_area_secs = std::min<uint64_t>(total, 4096);
    
    std::vector<std::string> notes;
    std::vector<std::unique_ptr<PartitionBlock>> parts;
    std::vector<Candidate> anchors;
    std::map<uint64_t, uint32_t> boots;
    
    struct RootNode { uint64_t sec; uint32_t spb; std::string label; };
    std::vector<RootNode> roots;
    
    uint64_t pos = 0;
    const uint64_t CHUNK_SECS = 8192;
    
    auto find_aligned = [](const std::vector<uint8_t>& chunk, const std::vector<uint8_t>& needle, std::function<void(uint32_t)> handler) {
        if (needle.empty() || chunk.size() < needle.size()) return;
        for (size_t i = 0; i <= chunk.size() - needle.size(); i += 512) {
            bool match = true;
            for (size_t j = 0; j < needle.size(); ++j) {
                if (chunk[i+j] != needle[j]) { match = false; break; }
            }
            if (match) handler(i);
        }
    };
    
    std::vector<uint8_t> pat_part = {'P','A','R','T'};
    std::vector<uint8_t> pat_dos = {'D','O','S'};
    std::vector<uint8_t> pat_sfs = {'S','F','S','\0'};
    std::vector<uint8_t> pat_pfs3 = {'P','F','S','\x03'};
    std::vector<uint8_t> pat_afs = {'A','F','S','\x01'};
    std::vector<uint8_t> pat_root = {0,0,0,2,0,0,0,0,0,0,0,0};
    
    while (pos < total) {
        uint64_t n = std::min<uint64_t>(CHUNK_SECS, total - pos);
        std::vector<uint8_t> chunk(n * 512);
        blkdev->read(pos * 512, chunk);
        
        find_aligned(chunk, pat_part, [&](uint32_t off) {
            uint64_t sec = pos + off / 512;
            if (sec < rdb_area_secs) {
                auto pb = std::make_unique<PartitionBlock>(blkdev, sec);
                if (pb->read()) parts.push_back(std::move(pb));
            }
        });
        
        find_aligned(chunk, pat_dos, [&](uint32_t off) {
            uint8_t flavor = chunk[off + 3];
            if (flavor <= 7) boots[pos + off / 512] = 0x444F5300 + flavor;
        });
        
        find_aligned(chunk, pat_sfs, [&](uint32_t off) {
            if (off + 512 > chunk.size()) return;
            uint16_t version = (chunk[off+12] << 8) | chunk[off+13];
            uint32_t ownblock = (chunk[off+8] << 24) | (chunk[off+9] << 16) | (chunk[off+10] << 8) | chunk[off+11];
            if (version == 3 && ownblock == 0) {
                uint32_t totalblocks = (chunk[off+48] << 24) | (chunk[off+49] << 16) | (chunk[off+50] << 8) | chunk[off+51];
                uint32_t blocksize = (chunk[off+52] << 24) | (chunk[off+53] << 16) | (chunk[off+54] << 8) | chunk[off+55];
                if (blocksize >= 512 && blocksize <= 32768 && totalblocks > 0) {
                    uint32_t spb = blocksize / 512;
                    Candidate c;
                    c.source = "sfs";
                    c.start_sec = pos + off / 512;
                    c.num_secs = totalblocks * spb;
                    c.dos_type = 0x53465300;
                    c.sec_per_blk = spb;
                    c.exact = true;
                    anchors.push_back(std::move(c));
                }
            }
        });
        
        auto on_pfs = [&](uint32_t off) {
            if (off + 512 > chunk.size()) return;
            uint32_t disksize = (chunk[off+84] << 24) | (chunk[off+85] << 16) | (chunk[off+86] << 8) | chunk[off+87];
            uint32_t options = (chunk[off+4] << 24) | (chunk[off+5] << 16) | (chunk[off+6] << 8) | chunk[off+7];
            uint64_t sec = pos + off / 512;
            if (disksize > 0 && (options & 1) && sec >= 2) {
                uint8_t nlen = chunk[off+20];
                std::string label((const char*)&chunk[off+21], std::min<uint8_t>(nlen, 31));
                Candidate c;
                c.source = "pfs3";
                c.start_sec = sec - 2;
                c.num_secs = disksize;
                c.dos_type = 0x50465303; // usually 03 for pfs3, we can just use 0x50465303
                c.sec_per_blk = 1; // wait, PFS3 doesn't rely on spb like this for geometry but default to 1
                c.label = label;
                c.exact = true;
                anchors.push_back(std::move(c));
            }
        };
        
        find_aligned(chunk, pat_pfs3, on_pfs);
        find_aligned(chunk, pat_afs, on_pfs);
        
        find_aligned(chunk, pat_root, [&](uint32_t off) {
            uint64_t sec = pos + off / 512;
            for (uint32_t spb : {1, 2, 4, 8}) {
                if (sec + spb > total) break;
                std::vector<uint8_t> raw(spb * 512);
                blkdev->read(sec * 512, raw);
                if (ffs_root_valid(raw, 512 * spb)) {
                    uint32_t bs = 512 * spb;
                    uint8_t nlen = raw[bs - 80];
                    std::string label((const char*)&raw[bs - 79], std::min<uint8_t>(nlen, 30));
                    roots.push_back({sec, spb, label});
                    break;
                }
            }
        });
        
        pos += n;
        if (progress) progress(pos, total);
    }
    
    std::vector<Candidate> ffs;
    std::vector<uint64_t> boot_list;
    for (auto const& [k, v] : boots) boot_list.push_back(k);
    
    for (size_t i = 0; i < boot_list.size(); ++i) {
        uint64_t bsec = boot_list[i];
        uint32_t flavor = boots[bsec];
        uint64_t nxt = (i + 1 < boot_list.size()) ? boot_list[i+1] : total;
        
        for (const auto& r : roots) {
            if (r.sec > bsec && r.sec < nxt) {
                uint64_t root_idx = (r.sec - bsec) / r.spb;
                if (root_idx < 2 || (r.sec - bsec) % r.spb != 0) continue;
                
                uint64_t tot = 2 * root_idx - 1;
                Candidate c;
                c.source = "ffs";
                c.start_sec = bsec;
                c.num_secs = tot * r.spb;
                c.dos_type = flavor;
                c.label = r.label;
                c.sec_per_blk = r.spb;
                c.exact = false;
                ffs.push_back(std::move(c));
                break;
            }
        }
    }
    
    std::vector<Candidate> cands;
    for (auto& pb : parts) {
        Candidate c;
        c.source = "part";
        c.start_sec = pb->envec.start_sec();
        c.num_secs = pb->envec.num_secs();
        c.dos_type = pb->envec.dos_type;
        c.label = pb->drive_name;
        c.sec_per_blk = std::max<uint32_t>(pb->envec.sec_per_blk, 1);
        c.part_blk = std::move(pb);
        c.exact = true;
        cands.push_back(std::move(c));
    }
    
    for (auto& a : anchors) {
        bool overlap = false;
        for (const auto& c : cands) {
            if (a.start_sec >= c.start_sec && a.start_sec < c.start_sec + c.num_secs) {
                overlap = true; break;
            }
        }
        if (!overlap) cands.push_back(std::move(a));
    }
    for (auto& a : ffs) {
        bool overlap = false;
        for (const auto& c : cands) {
            if (a.start_sec >= c.start_sec && a.start_sec < c.start_sec + c.num_secs) {
                overlap = true; break;
            }
        }
        if (!overlap) cands.push_back(std::move(a));
    }
    
    std::sort(cands.begin(), cands.end(), [](const Candidate& a, const Candidate& b) {
        return a.start_sec < b.start_sec;
    });
    
    std::vector<Candidate> pruned;
    for (auto& c : cands) {
        if (!pruned.empty() && c.start_sec < pruned.back().start_sec + pruned.back().num_secs) {
            auto& prev = pruned.back();
            if (c.exact && !prev.exact) {
                pruned.back() = std::move(c);
            } else {
                notes.push_back("dropped overlapping candidate: " + c.describe());
            }
            continue;
        }
        pruned.push_back(std::move(c));
    }
    
    if (pruned.empty()) {
        notes.push_back("no partition evidence found");
    }
    
    return {std::move(pruned), notes};
}

static std::pair<uint32_t, uint32_t> pick_geometry(const std::vector<Candidate>& candidates, uint64_t total_secs) {
    std::vector<std::pair<uint32_t, uint32_t>> options;
    uint32_t heads_arr[] = {16, 8, 4, 2, 1, 32, 64, 128, 255};
    uint32_t sectors_arr[] = {63, 32, 127, 126, 16};
    
    for (uint32_t heads : heads_arr) {
        for (uint32_t sectors : sectors_arr) {
            uint32_t cyl = heads * sectors;
            if (total_secs / cyl > 65535 || total_secs / cyl < 3) continue;
            
            bool ok = true;
            for (const auto& c : candidates) {
                if (c.start_sec % cyl != 0) { ok = false; break; }
            }
            if (ok) options.push_back({heads, sectors});
        }
    }
    if (!options.empty()) return options[0];
    return {1, 1};
}

std::unique_ptr<RDisk> Rescue::rebuild(std::shared_ptr<BlockDevice> blkdev, const std::vector<Candidate>& candidates, const std::string& drv_prefix) {
    if (blkdev->is_read_only()) throw std::runtime_error("device is read-only");
    if (candidates.empty()) throw std::runtime_error("nothing to rebuild");
    
    bool reuse = true;
    for (const auto& c : candidates) {
        if (!c.part_blk) { reuse = false; break; }
    }
    
    uint32_t heads, sectors;
    if (reuse) {
        heads = candidates[0].part_blk->envec.surfaces;
        sectors = candidates[0].part_blk->envec.blk_per_trk;
    } else {
        auto geom = pick_geometry(candidates, blkdev->num_blocks());
        heads = geom.first;
        sectors = geom.second;
    }
    
    uint64_t first_start = candidates[0].start_sec;
    for (const auto& c : candidates) {
        if (c.start_sec < first_start) first_start = c.start_sec;
    }
    
    uint32_t cyl = heads * sectors;
    uint32_t rdb_cyls = std::max<uint32_t>(1, first_start / cyl);
    if (first_start >= cyl) rdb_cyls = std::min<uint32_t>(rdb_cyls, 2);
    else rdb_cyls = 1;
    
    auto rd = RDisk::create(blkdev, sectors, heads, rdb_cyls);
    
    if (reuse) {
        for (size_t i = 0; i < candidates.size(); ++i) {
            auto& pb = candidates[i].part_blk;
            pb->next = (i + 1 < candidates.size()) ? candidates[i+1].part_blk->blk_num() : 0xFFFFFFFF;
            pb->write();
            if (pb->blk_num() > rd->rdsk_blk().high_rdsk_block) 
                const_cast<RDBlock&>(rd->rdsk_blk()).high_rdsk_block = pb->blk_num();
        }
        const_cast<RDBlock&>(rd->rdsk_blk()).partition_list = candidates[0].part_blk->blk_num();
        const_cast<RDBlock&>(rd->rdsk_blk()).write();
        
        // RDisk::create already sets up the structures, but since we manually linked the partition list
        // we might want to reload it, but just returning `rd` is fine if the user calls `RDisk::open` next.
        // Actually we can just reload it into a fresh instance.
        return RDisk::open(blkdev);
    }
    
    for (size_t i = 0; i < candidates.size(); ++i) {
        const auto& c = candidates[i];
        uint32_t low = c.start_sec / cyl;
        uint32_t high = ((c.start_sec + c.num_secs) + cyl - 1) / cyl - 1;
        if (high > rd->rdsk_blk().hi_cylinder) high = rd->rdsk_blk().hi_cylinder;
        
        if (i + 1 < candidates.size()) {
            uint32_t next_low = candidates[i+1].start_sec / cyl;
            if (next_low > 0 && high > next_low - 1) {
                high = next_low - 1;
            }
        }
        
        std::string name = (c.source == "part" && !c.label.empty()) ? c.label : (drv_prefix + std::to_string(i));
        rd->add_partition_exact(name, low, high, c.dos_type, c.sec_per_blk);
    }
    
    return rd;
}

} // namespace amidisk

File: amidisk_python/src/amidisk/rdb/rescue.py
"""RDB rescue: analyze a disk whose RDB was overwritten and rebuild it.

Scenario: the RDSK block (and possibly the PART chain) at the start of
the disk got trashed -- by a PC partitioner, an accidental dd, a format
of the wrong drive -- but the partition *contents* (the filesystems)
are still on disk. This module finds them again:

  evidence, best first:
  1. orphaned PART blocks in the old RDB area -- exact geometry
  2. SFS root blocks   -- live at partition start, carry exact size
  3. PFS3 root blocks  -- live at partition start + 2, carry exact size
  4. FFS boot + root block pairs -- root sits at the middle of the
     partition, so its distance from the boot block gives the size

The rebuilder then writes a fresh RDSK (reusing intact PART blocks when
found, synthesizing them otherwise), after backing up the old RDB area
to a file.
"""

import struct

from .blocks import PartitionBlock, dos_type_to_str
from .rdisk import RDisk, RDiskError, END

CHUNK_SECS = 8192  # 4 MB scan chunks


class Candidate:
    """One recovered partition candidate (all units: 512-byte sectors)."""

    def __init__(self, source, start, num_secs, dos_type, label=None,
                 sec_per_blk=1, part_blk=None, exact=True):
        self.source = source          # 'part' | 'sfs' | 'pfs3' | 'ffs'
        self.start = start
        self.num_secs = num_secs      # may be approximate for ffs
        self.dos_type = dos_type
        self.label = label
        self.sec_per_blk = sec_per_blk
        self.part_blk = part_blk      # orphaned PartitionBlock, if any
        self.exact = exact

    def end(self):
        return self.start + self.num_secs

    def describe(self):
        return "%-5s @sector %-9d %8.1f MB  %-7s %s%s" % (
            self.source, self.start, self.num_secs * 512 / 1e6,
            dos_type_to_str(self.dos_type),
            repr(self.label) if self.label else "-",
            "" if self.exact else "  (size estimated)",
        )


def _ffs_root_valid(raw, bs):
    """Is `raw[:bs]` a valid FFS root block?"""
    if len(raw) < bs:
        return False
    nl = bs // 4
    if struct.unpack_from(">I", raw, 0)[0] != 2:
        return False
    if struct.unpack_from(">i", raw, bs - 4)[0] != 1:
        return False
    s = 0
    for (v,) in struct.iter_unpack(">I", raw[:bs]):
        s = (s + v) & 0xFFFFFFFF
    return s == 0


def _sfs_checksum_ok(raw):
    s = 0
    for (v,) in struct.iter_unpack(">I", raw):
        s = (s + v) & 0xFFFFFFFF
    return s == 0xFFFFFFFF


def scan(blkdev, progress=None, rdb_area_secs=None):
    """Scan the device for partition evidence. Returns (candidates, notes)."""
    total = blkdev.num_blocks
    if rdb_area_secs is None:
        rdb_area_secs = min(total, 4096)
    notes = []
    parts = []
    anchors = []
    boots = {}  # sector -> dostype (FFS boot block candidates)

    roots = []  # (sector, spb, label) of validated FFS root blocks

    def find_aligned(chunk, needle, handler):
        i = chunk.find(needle)
        while i != -1:
            if i % 512 == 0:
                handler(i)
            i = chunk.find(needle, i + 1)

    # prefilter for an FFS root block: type=2, header_key=0, high_seq=0
    ROOT_PAT = b"\x00\x00\x00\x02" + b"\x00" * 8

    pos = 0
    while pos < total:
        n = min(CHUNK_SECS, total - pos)
        chunk = blkdev.read(pos, n)

        def on_part(off, pos=pos):
            sec = pos + off // 512
            if sec < rdb_area_secs:
                pb = PartitionBlock(blkdev, sec)
                if pb.read():
                    parts.append(pb)

        def on_dos(off, chunk=chunk, pos=pos):
            flavor = chunk[off + 3]
            if flavor <= 7:
                boots[pos + off // 512] = 0x444F5300 + flavor

        def on_sfs(off, chunk=chunk, pos=pos):
            raw = chunk[off : off + 512]
            if len(raw) < 512:
                return
            version = struct.unpack_from(">H", raw, 12)[0]
            ownblock = struct.unpack_from(">I", raw, 8)[0]
            if version == 3 and ownblock == 0:
                totalblocks, blocksize = struct.unpack_from(">2I", raw, 48)
                if 512 <= blocksize <= 32768 and totalblocks > 0:
                    spb = blocksize // 512
                    anchors.append(
                        Candidate(
                            "sfs", pos + off // 512, totalblocks * spb,
                            0x53465300, sec_per_blk=spb,
                        )
                    )

        def on_pfs(off, chunk=chunk, pos=pos):
            raw = chunk[off : off + 512]
            if len(raw) < 512:
                return
            disksize = struct.unpack_from(">I", raw, 84)[0]
            options = struct.unpack_from(">I", raw, 4)[0]
            sec = pos + off // 512
            if disksize and (options & 1) and sec >= 2:  # MODE_HARDDISK
                nlen = raw[20]
                label = raw[21 : 21 + min(nlen, 31)].decode("latin-1", "replace")
                anchors.append(
                    Candidate("pfs3", sec - 2, disksize, 0x50465303, label=label)
                )

        def on_root(off, pos=pos):
            sec = pos + off // 512
            for spb in (1, 2, 4, 8):
                if sec + spb > total:
                    break
                raw = blkdev.read(sec, spb)
                if _ffs_root_valid(raw, 512 * spb):
                    bs = 512 * spb
                    nlen = raw[bs - 80]
                    label = raw[bs - 79 : bs - 79 + min(nlen, 30)].decode(
                        "latin-1", "replace"
                    )
                    roots.append((sec, spb, label))
                    break

        find_aligned(chunk, b"PART", on_part)
        find_aligned(chunk, b"DOS", on_dos)
        find_aligned(chunk, b"SFS\x00", on_sfs)
        find_aligned(chunk, b"PFS\x01", on_pfs)
        find_aligned(chunk, b"AFS\x01", on_pfs)
        find_aligned(chunk, ROOT_PAT, on_root)
        pos += n
        if progress:
            progress(pos, total)

    # pair FFS boot blocks with their root blocks: the root block sits at
    # fs-block (2 + total - 1) // 2, so total ~= 2 * (root - boot) / spb
    ffs = []
    boot_list = sorted(boots)
    for i, bsec in enumerate(boot_list):
        flavor = boots[bsec]
        nxt = boot_list[i + 1] if i + 1 < len(boot_list) else total
        for rsec, spb, label in roots:
            if not (bsec < rsec < nxt):
                continue
            root_idx = (rsec - bsec) // spb
            if root_idx < 2 or (rsec - bsec) % spb:
                continue
            # (2 + tot - 1) // 2 == root_idx admits tot and tot+1; use the
            # smaller -- the rebuilder rounds the end up to the cylinder
            # boundary, which recovers the exact original size, while the
            # larger estimate would spill one block into the next cylinder
            tot = 2 * root_idx - 1
            ffs.append(
                Candidate(
                    "ffs", bsec, tot * spb, flavor,
                    label=label, sec_per_blk=spb, exact=False,
                )
            )
            break

    # orphaned PART blocks win over anchors covering the same region
    cands = []
    for pb in parts:
        env = pb.dos_env
        cands.append(
            Candidate(
                "part", env.start_sec(), env.num_secs(), env.dos_type,
                label=pb.drv_name, sec_per_blk=max(env.sec_per_blk, 1),
                part_blk=pb,
            )
        )
    for a in anchors + ffs:
        if not any(c.start <= a.start < c.end() for c in cands):
            cands.append(a)
    cands.sort(key=lambda c: c.start)

    # drop overlapping duplicates (keep the earlier, more exact one)
    pruned = []
    for c in cands:
        if pruned and c.start < pruned[-1].end():
            prev = pruned[-1]
            if c.exact and not prev.exact:
                pruned[-1] = c
            else:
                notes.append(
                    "dropped overlapping candidate: %s" % c.describe()
                )
            continue
        pruned.append(c)
    if not pruned:
        notes.append("no partition evidence found")
    return pruned, notes


def _pick_geometry(candidates, total_secs):
    """Choose heads/sectors such that all candidate starts are
    cylinder-aligned and the cylinder count fits in 16 bits."""
    starts = [c.start for c in candidates]
    options = []
    for heads in (16, 8, 4, 2, 1, 32, 64, 128, 255):
        for sectors in (63, 32, 127, 126, 16):
            cyl = heads * sectors
            if total_secs // cyl > 65535 or total_secs // cyl < 3:
                continue
            if all(s % cyl == 0 for s in starts):
                options.append((heads, sectors))
    if options:
        return options[0]
    # last resort: 1 sector per cylinder aligns with anything
    return 1, 1


def rebuild(blkdev, candidates, backup_path=None, drv_prefix="DH"):
    """Write a fresh RDB describing `candidates`. Returns the new RDisk."""
    if blkdev.read_only:
        raise RDiskError("device is read-only")
    if not candidates:
        raise RDiskError("nothing to rebuild")

    if backup_path:
        area = min(blkdev.num_blocks, 2048)
        with open(backup_path, "wb") as fh:
            fh.write(blkdev.read(0, area))

    reuse = all(c.part_blk is not None for c in candidates)
    if reuse:
        env = candidates[0].part_blk.dos_env
        heads, sectors = env.surfaces, env.blk_per_trk
    else:
        heads, sectors = _pick_geometry(candidates, blkdev.num_blocks)

    first_start = min(c.start for c in candidates)
    cyl = heads * sectors
    rdb_cyls = max(1, first_start // cyl)
    rdb_cyls = min(rdb_cyls, 2) if first_start >= cyl else 1
    rd = RDisk.create(blkdev, sectors=sectors, heads=heads, rdb_cyls=rdb_cyls)

    if reuse:
        # relink the surviving PART blocks under the fresh RDSK
        ordered = sorted(candidates, key=lambda c: c.start)
        for i, c in enumerate(ordered):
            pb = c.part_blk
            pb.next = ordered[i + 1].part_blk.blk_num if i + 1 < len(ordered) else END
            pb.write()
            rd.rdb.high_rdsk_block = max(rd.rdb.high_rdsk_block, pb.blk_num)
        rd.rdb.partition_list = ordered[0].part_blk.blk_num
        rd._write_rdb()
        rd.open()
        return rd

    ordered = sorted(candidates, key=lambda c: c.start)
    for i, c in enumerate(ordered):
        low = c.start // cyl
        high = (c.end() + cyl - 1) // cyl - 1
        high = min(high, rd.rdb.hi_cylinder)
        if i + 1 < len(ordered):  # never spill into the next partition
            high = min(high, ordered[i + 1].start // cyl - 1)
        name = c.label if (c.label and c.source == "part") else "%s%d" % (drv_prefix, i)
        rd.add_partition(
            name,
            num_cyls=None,
            size_bytes=None,
            low_cyl=low,
            high_cyl=high,
            dos_type=c.dos_type,
            sec_per_blk=c.sec_per_blk,
        )
    return rd
